### PR TITLE
Rewriting image augmenter  

### PR DESCRIPTION
**Pull request by chsasank**
Hi,

As discussed on #2394, I rewrote `ImageDataGenerator` using homographies. There's a significant speedup compared to the original implementation.

Minor downside is that scikit-image is now a dependency.

### CODE REVIEW

**Comment by joelthchao** [comment]
I think we should keep the old one and create new class `SkimageDataGenerator(ImageDataGenerator)` for better support.

**Comment by chsasank** [comment]
Actually, the old one is buggy.
I ran a small [script](https://gist.github.com/chsasank/2f392206b07b70a298b93f518bdb5b29) and saved images generated by the augmenter.

Original image (`random_img.jpg`):
random_img


These are generated by keras:

gen_0_keras
gen_1_keras
gen_2_keras
gen_3_keras
gen_4_keras
 
And these by the fork:
gen_0_fork
gen_1_fork
gen_2_fork
gen_3_fork
gen_4_fork

So, this warrants the removal of old `ImageDataGenerator`

**Comment by albertomontesg** [comment]
I think that the transformation methods shouldn't be removed because it would be useful to use them at writing your own DataGenerator with custom data augmentation.

**Comment by joelthchao** [comment]
@albertomontesg Thanks, these useful methods should be kept for the future use.

**Comment by chsasank** [comment]
> I think that the transformation methods shouldn't be removed because it would be useful to use them at writing your own DataGenerator with custom data augmentation.

But people can do the same with the new one, right?
All that has to be changed is `random_transform`

**Comment by joelthchao** [comment]
@chsasank Better way is to make functions which only do a single transformation. By the way, I fix several bugs in `fit` and I am going to temporally keep those codes and wait for next refactor to replace them.

**Comment by chsasank** [comment]
@joelthchao  
>Better way is to make functions which only do a single transformation

But this defeats the purpose of this PR right?

What bugs were there in `fit`?

**Comment by joelthchao** [comment]
@chsasank `fit` convert numpy array to PIL Image and send to `random_transform`. Your code assume input is in numpy type, which is right during `flow` but fail in `fit` when it calls `x.dtype`.

**Comment by chsasank** [comment]
Cool, I get what you mean by keeping transformation methods intact.
I'll rewrite these functions to use skimage instead of ndimage.

**Comment by fchollet** [comment]
I don't think we should depend on skimage at all. Any way to get the same speed-up while only relying on Numpy/Scipy?

**Comment by dansbecker** [comment]
Wow.  That issue with the colors seems bad.  If you guys want someone to look into that, I'm happy to do so.  If we're pulling all that code out anyway, then I won't spend time on it.

**Comment by chsasank** [comment]
Hi @fchollet ,
I think we can use [this](http://docs.scipy.org/doc/scipy-0.16.0/reference/generated/scipy.ndimage.interpolation.affine_transform.html#scipy.ndimage.interpolation.affine_transform) instead of skimage. 
I'll test and make changes.

**Comment by fchollet** [comment]
> Wow. That issue with the colors seems bad.

I'm skeptical this is for real, I remember generating test images when this code was initially written and they looked fine. I'm 95% sure that OP is either normalizing the arrays in some way, or messing up the image saving step.

**Comment by fchollet** [comment]
> I think we can use this instead of skimage. 

Cool, that sounds good.

**Comment by chsasank** [comment]
I have rewritten using scipy's ndimage.
I think the bug in the previous code was here [1](https://github.com/fchollet/keras/blob/master/keras/preprocessing/image.py#L37) and [2](https://github.com/fchollet/keras/blob/master/keras/preprocessing/image.py#L61).

ndimage treats every dimension in an array as spatial dimension. See [this](http://stackoverflow.com/a/9284454/6212301).

So the operation [1] would work only if channel axis = 0 while [2] would work only if channel axis = 2.
I have fixed this.

**Comment by chsasank** [comment]
BTW, can somebody explain me what `random_channel_shift` is?
There's a `todo` in the code.

I will try to implement it.

**Comment by fchollet** [comment]
> ndimage treats every dimension in an array as spatial dimension.

Right, I get it now.

> BTW, can somebody explain me what random_channel_shift is?

It's a random shift in RGB color channels, effectively changing the color scheme of a picture. e.g. +10 on the R channel, -20 on the G channel, +5 on the B channel...

**Comment by chsasank** [comment]
Is this like what alexnet paper does?

From the paper:
>The second form of data augmentation consists of altering the intensities of the RGB channels in training images. Specifically, we perform PCA on the set of RGB pixel values throughout the ImageNet training set. To each training image, we add multiples of the found principal components, with magnitudes proportional to the corresponding eigenvalues times a random variable drawn from
a Gaussian with mean zero and standard deviation 0.1

**Comment by fchollet** [comment]
> Is this like what alexnet paper does?

It may be overkill, and it is less practical because it requires fitting on training data, but yes, that would be a good strategy.

**Comment by chsasank** [comment]
We can have pertained components trained on imagenet as an option if an user doesn't want to fit on his own data. Is this a good idea?

**Comment by chsasank** [comment]
Also, `featurewise_centre` and `featurewise_centre`  do not do simple normalisation of channels but they have mean image and stddev image. Is this intended?

**Comment by joelthchao** [comment]
@chsasank 
> We can have pertained components trained on imagenet as an option if an user doesn't want to fit on his own data. Is this a good idea?

I think it would make this component too complicated to use, probably you can make a comprehensive example to illustrate how to do proper data augmentation based on Imagenet

> Also, featurewise_centre and featurewise_centre do not do simple normalisation of channels but they have mean image and stddev image. Is this intended?

It fits to train data, which is a standard normalization method in computer vision.

Also, you can add `random_zoom` to TODO or I can implement it now.

**Comment by fchollet** [comment]
> Also, featurewise_centre and featurewise_centre do not do simple normalisation of channels but they have mean image and stddev image. Is this intended?

Channelwise normalization would be a useful addition as well. I think it would be fine to make that mode work on separate channels differently rather than globally on all channels.

Please let me know when this is ready for merging.

**Comment by joelthchao** [comment]
Currently, `standardize` do the following things:
1. `samplewise`, normalize over channel for a single image
2. `featurewise`: substract mean image and divide std image of data set, which is channel independent.
@fchollet do you mean creating a `channelwise` mode? 
PS. most of works only do `featurewise` normalization

File: keras/preprocessing/image.py [hunk 1]
```diff
@@ -70,9 +70,13 @@ def random_barrel_transform(x, intensity):
     # TODO
     pass
 
-def random_channel_shift(x, rg):
-    # TODO
-    pass
+def random_channel_shift(x, intensity, channel_index=0):
+    x = np.rollaxis(x, channel_index, 0)
+    channel_images = [np.clip(x_channel + np.random.uniform(-intensity, intensity), 0, 1)
```
Comment (chsasank) [inline keras/preprocessing/image.py]: Why clip to [0,1]?
There's no guarantee that X passed to the datagenerator will be within 0 and 1.

File: keras/preprocessing/image.py [hunk 2]
```diff
@@ -6,105 +6,137 @@
...
-    shear_matrix = np.array([[1.0, -math.sin(shear), 0.0],
-                            [0.0, math.cos(shear), 0.0],
-                            [0.0, 0.0, 1.0]])
-    x = ndimage.interpolation.affine_transform(x, shear_matrix,
-                                               mode=fill_mode,
-                                               order=3,
-                                               cval=cval)
+def random_channel_shift(x, intensity, channel_index=0):
+    x = np.rollaxis(x, channel_index, 0)
+    channel_images = [np.clip(x_channel + np.random.uniform(-intensity, intensity), 0, 1)
```
Comment (chsasank) [inline keras/preprocessing/image.py]: Why clip to [0,1]
There's no guarantee that X passed to data generator will be within 0 and 1.
For example uint8 images can be passed to data generator.

Comment (joelthchao) [inline keras/preprocessing/image.py]: Got it! Let me check if other codes make such guarantee.

Comment (joelthchao) [inline keras/preprocessing/image.py]: array_to_img need argument scale to decide if it is needed to scale to uint8. And the previous code assume X is always within 0 and 1. We may need to fix it.

Comment (chsasank) [inline keras/preprocessing/image.py]: >And the previous code assume X is always within 0 and 1. We may need to fix it.

No all the random transformations do not assume anything about the scale. I have checked it passing uint8 images.

Also, `array_to_img` has an argument `scale` to rescale the images to [0,256] for writing purposes.

Comment (chsasank) [inline keras/preprocessing/image.py]: For `random_channel_shift` we can change the offending line to 
```
channel_images = [np.clip(x_channel + np.random.uniform(-intensity, intensity), np.min(x), np.max(x)) 
                  for x_channel in x]
```

**Comment by joelthchao** [comment]
@fchollet I think it is ready. @chsasank can you double check the codes?
Barrel/Fisheye effect may need cv2 library support, so I will leave them along.

File: docs/templates/preprocessing/image.md [hunk 3]
```diff
@@ -28,6 +28,10 @@ Generate batches of tensor image data with real-time data augmentation. The data
     - __width_shift_range__: Float (fraction of total width). Range for random horizontal shifts.
     - __height_shift_range__: Float (fraction of total height). Range for random vertical shifts.
     - __shear_range__: Float. Shear Intensity (Shear angle in counter-clockwise direction as radians)
+    - __zoom_range__: Float or [lower, upper]. Range for random zoom.
```
Comment (chsasank) [inline docs/templates/preprocessing/image.md]: Shouldn't we explain if zoom_range is float `[lower, upper]` is `[1-zoom_range, 1+ zoom_range]`, because it is not apparent from the documentation.

**Comment by chsasank** [comment]
>Barrel/Fisheye effect may need cv2 library support, so I will leave them alone.

We might be able to implement it with `ndimages`, but it might be slow. See [this](http://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.ndimage.interpolation.map_coordinates.html).
We can leave this for a future PR.

File: keras/preprocessing/image.py [hunk 4]
```diff
@@ -6,105 +6,137 @@
...
-                            [0.0, math.cos(shear), 0.0],
-                            [0.0, 0.0, 1.0]])
-    x = ndimage.interpolation.affine_transform(x, shear_matrix,
-                                               mode=fill_mode,
-                                               order=3,
-                                               cval=cval)
+def random_channel_shift(x, intensity, channel_index=0):
+    x = np.rollaxis(x, channel_index, 0)
+    min_x, max_x = np.min(x), np.max(x)
+    channel_images = [np.clip(x_channel + np.random.uniform(-intensity, intensity), min_x, max_x)
```
Comment (chsasank) [inline keras/preprocessing/image.py]: On a second thought, for better UX shouldn't we ask user to specify channel shift `intensity` in terms of fraction of  dynamic range, instead of absolute values? (like in `height_shift_range `)

Comment (joelthchao) [inline keras/preprocessing/image.py]: If fraction is used for channel shift `intensity` then we might need to know whether user is using `uint8` or `float32`. But can we know that?

Comment (chsasank) [inline keras/preprocessing/image.py]: We can just use `min_x, max_x = np.min(x), np.max(x)` for dynamic range needs right?

Comment (joelthchao) [inline keras/preprocessing/image.py]: That is not 100% guarantee and it will result in different intensity for different image.

Comment (chsasank) [inline keras/preprocessing/image.py]: True, it might be slightly off [0,255] for uint8 image or [0,1] for a float image, but it usually is a very good approximation. 
>it will result in different intensity for different image.

We're adding random intensities anyway, right?

Comment (joelthchao) [inline keras/preprocessing/image.py]: For example, given 20% intensity shift, an image with pixel value [50, 100] will have larger shift than an image with pixel value in [70, 80].

Comment (chsasank) [inline keras/preprocessing/image.py]: Of course. But natural images usually use almost complete dynamic range. Think about it, in a uint8 64x64x3 image, there are about 12k pixels. There's a very high probability min is 0 and max is 255 or 254. 

It's just a suggestion. We can keep current one as it is, if you think that is better.

**Comment by fchollet** [comment]
> @fchollet I think it is ready.

Cool, will review later today. Thanks!

File: keras/preprocessing/image.py [hunk 5]
```diff
@@ -6,105 +6,137 @@
...
+from six.moves import range
 from os import listdir
 from os.path import isfile, join
-import math
-from six.moves import range
 import threading
 
+def random_rotation(x, rg, row_index=1, col_index=2, channel_index=0,
+                    fill_mode='nearest', cval=0.):
+    theta = np.pi/180*np.random.uniform(-rg, rg)
```
Comment (fchollet) [inline keras/preprocessing/image.py]: Style nitpick, but please insert spaces around operators (`*`, `+`, etc).

File: keras/preprocessing/image.py [hunk 6]
```diff
@@ -6,105 +6,137 @@
...
-    x = x[::-1, ...]
-    x = x.swapaxes(0, axis)
+    h, w = x.shape[row_index], x.shape[col_index]
+    transform_matrix = transform_matrix_offset_center(shear_matrix, h, w)
+    x = apply_transform(x, transform_matrix, channel_index, fill_mode, cval)
     return x
 
+def random_zoom(x, zoom_range, row_index=1, col_index=2, channel_index=0,
+                fill_mode='nearest', cval=0.):
+    if zoom_range != [1., 1.]:
```
Comment (fchollet) [inline keras/preprocessing/image.py]: I think we should expect a list, but also allow for tuples. In which case this comparison won't work; better to compare each member: `zoom_range[0] == 1 and zoom_range[1] == 1` .

Also, since we make the assumption that `zoom_range` is going to be a 2-element list/tuple, better check first and raise a helpful error message if it's not the case.

File: keras/preprocessing/image.py [hunk 7]
```diff
@@ -6,105 +6,137 @@
...
+    x = np.rollaxis(x, channel_index, 0)
+    min_x, max_x = np.min(x), np.max(x)
+    channel_images = [np.clip(x_channel + np.random.uniform(-intensity, intensity), min_x, max_x)
+                      for x_channel in x]
+    x = np.stack(channel_images, axis=0)
+    x = np.rollaxis(x, 0, channel_index+1)
     return x
 
+def transform_matrix_offset_center(matrix, x, y):
+    o_x = float(x)/2 + 0.5
```
Comment (fchollet) [inline keras/preprocessing/image.py]: Spaces around `/`

**Comment by fchollet** [comment]
Cool, looks good to me. Just a few minor nitpicks.

**Comment by fchollet** [comment]
Thank you for this useful contribution, merging now.
